### ISSUE TITLE
parse_single_example() should give dense array for VarLenFeature

### ISSUE THREAD

**Issue by phsmit** [tensorflow/tensorflow#976]
On master, the signature of parse_single_example has changed. When testing I noticed something unexpected; a tf.VarLenFeature results in a SparseTensor instead of a dense one.

    features = tf.parse_single_example(
      serialized_example,
      features = {
        'data': tf.VarLenFeature(tf.float32),
      }
    )

`features['data']` is now a SparseTensor. I don't see really why, shouldn't this be a (standard dense) Tensor instead?

**Comment by ebrevdo** [comment]
If you then want to pass this data through, e.g., a FIFOQueue, so you can then perform minibatch processing, you will have different shaped tensors going in and as a result will not be able to dequeue_many.

Instead, you want to serialize these 1D sparse tensors via sparse_ops.serialize_sparse, and feed the resulting string into the FIFOQueue.  Then after a dequeue_many, you'll use sparse_ops.deserialize_many_sparse to get back a 2D SparseTensor with the dimensions [batch index, variable depth].

**Comment by ebrevdo** [comment]
In other words, works as intended.

**Comment by phsmit** [comment]
I can't seem to find the ops you are mentioning in the documentation, but I'm sure what you are mentioning is a valid use case.

My use case is a bit different though. My code looks at the moment something like this:

    FEA_DIM=13
    f = parse_single_example(
      se,
      features={
        'input': tf.VarLenFeature(tf.float32),
        'output': tf.VarLenFeature(tf.int64),
      }
    )
    
    x = tf.sparse_tensor_to_dense(features['input'])
    y = tf.sparse_tensor_to_dense(features['output'])
    
    x = tf.reshape(x, [-1,FEA_DIM])
    y = tf.reshape(y, [-1,1])

    return tf.train.shuffle_batch([x,y], enqueue_many=True)

So I use a single TfRecord to hold multiple x-y pairs.* For me it looks wrong that data that is stored densely (as a protobuf float_list), would come to a sparse array first, when I am going to transform it right back to a dense array. Note that I can't use tf.FixedLenSequenceFeature as I have no clue how many examples there will be in one record (not even a maximum).



* In reality this is timeseries (speech) data and I do some striding before the features to the queue. This is why there are multiple pairs, it saves space as I don't have to store the context on disk for every x-y separately.

**Comment by ebrevdo** [comment]
Aside from the bad naming, tf.FixedLenSequenceFeature is exactly what you want.  The "FixedLen" in your case is 13, but the first dimension of SequenceFeatures is considered unknown.  Store your data in SequenceExample proto (which is meant for data like speech), use parse_single_sequence_example, and pull out FixedLenSequenceFeature(float32, FEA_DIM),  and FixedLenSequenceFeature(int64, 1).  This will immediately return to you matrices of the shape [-1, FEA_DIM] and [-1, 1].

Alternatively you can just use tf.reshape(x.values, [-1, FEA_DIM]) and tf.reshape(y.values, [-1, 1])  since you can access the values vectors directly from a SparseTensor via property values.  But this won't scale if you start having more complicated features (e.g. 2D per time step, variable-length per time step).

The former solution

**Comment by ebrevdo** [comment]
... the former solution will allow you to declare the more complex input formats and is meant for speech and other sequential data of unknown # of frames.

**Comment by phsmit** [comment]
Thanks! I have now changed my code to use the SequenceFeatures with parse_single_sequence_example and it works like a charm!

**Comment by jinghuangzhu** [comment]
Glad to find this discussion because I am learning to use tensorflow to deal with speech data, and never saw any documentation on parse_single_sequence_example and SequenceExample proto. Seems TF hides lots of good stuff undernearth :). Can you please provide an example code to use these on speech data? Thanks a lot.

**Comment by chenghuige** [comment]
@jinghuangzhu  for anyone intersted in using sequence example, here is a good example
https://github.com/tensorflow/models/blob/master/im2txt/im2txt/ops/inputs.py
https://github.com/tensorflow/models/blob/master/im2txt/im2txt/show_and_tell_model.py

**Comment by songmeixu** [comment]
@ebrevd

**Comment by songmeixu** [comment]
@ebrevdo I do what you said. 
```python
FEAT_DIM=4
context, seq_features = tf.parse_single_sequence_example(
    file_example,
    context_features={
        'numFrames': tf.FixedLenFeature([], dtype=tf.int64),
    },
    sequence_features={
        'feats': tf.FixedLenSequenceFeature([FEAT_DIM], dtype=tf.float32),
        'labels': tf.FixedLenSequenceFeature([1], dtype=tf.int64),
    }
)
```
But a error happend:
>InvalidArgumentError (see above for traceback): Name: , Key: feats, Index: 0.  Number of float values != expected.  values size: 8 but output shape: [4]

PS: the matrix to be parsed is [2,4]
Can you give a workable sample?

**Comment by 345ishaan** [comment]
@ebrevdo  and @psmit 

I am trying to parse the image and the corresponding bounding box locations corresponding to number of objects in the image through a SequenceExample Proto.

But I am not able to dequeue the image and the corresponding bounding box locations.

Here is my implementation:-

def load_from_tfRecord(self,filename_queue):
		
		reader = tf.TFRecordReader()
		_, serialized_example = reader.read(filename_queue)

		context,sequence = tf.parse_single_sequence_example(
					serialized_example,
					context_features={'image_raw':tf.FixedLenFeature([], tf.string)},
					sequence_features={'locations':tf.FixedLenSequenceFeature([], dtype=tf.float32)}
					)
		
		image = tf.decode_raw(context['image_raw'], tf.uint8)
		
		image_shape = tf.stack([1,self.ip_width,self.ip_height,3])
		image = tf.cast(tf.reshape(image,image_shape),tf.float32)
		locations = sequence['locations']
		batch_images = tf.train.shuffle_batch([image]
			,enqueue_many=True,batch_size=self.batch_size,num_threads=1,capacity=1000,min_after_dequeue=500)

		return batch_images

**Comment by 345ishaan** [comment]
@ebrevdo  and @psmit 

I am trying to parse SequenceExamples where my context feature is an image and the features are the bounding box locations for the objects in the image.

I am not able to dequeue the data.
**Here is how I create the Example:-**

context = tf.train.Features(feature={'image_raw':self._bytes_feature(image)})
			feature_lists = tf.train.FeatureLists(feature_list={'locations':self._float_feature_list(list(locs[:,0]))})

			example = tf.train.SequenceExample(context=context,feature_lists=feature_lists)

**Here is how I am enqueing:-**

def load_from_tfRecord(self,filename_queue):
		
		reader = tf.TFRecordReader()
		_, serialized_example = reader.read(filename_queue)

		context,sequence = tf.parse_single_sequence_example(
					serialized_example,
					context_features={'image_raw':tf.FixedLenFeature([], tf.string)},
					sequence_features={'locations':tf.FixedLenSequenceFeature([], dtype=tf.float32)}
					)
		
		image = tf.decode_raw(context['image_raw'], tf.uint8)
		#locations = tf.decode_raw(features['locations'], tf.int32)

		image_shape = tf.stack([1,self.ip_width,self.ip_height,3])
		image = tf.cast(tf.reshape(image,image_shape),tf.float32)

		# loc_shape = tf.stack([n_locs,4])
		# locations = tf.reshape(locations,loc_shape)

		locations = sequence['locations']
		


		batch_images,batch_locations = tf.train.shuffle_batch([image,locations]
			,enqueue_many=True,batch_size=self.batch_size,num_threads=1,capacity=1000,min_after_dequeue=500)

		return batch_images,batch_locations

**Here is how I am dequeuing:-**

print "Loading Trained Model"
		self.session.run(tf.local_variables_initializer())

		#self.session.run(tf.group(tf.global_variables_initializer(),tf.local_variables_initializer()))

		coord = tf.train.Coordinator()
		threads = tf.train.start_queue_runners(sess=self.session,coord=coord)

		try:
			while not coord.should_stop():
				batch_imgs,batch_labels = self.session.run([self.images,self.labels])
				print batch_imgs.shape,batch_labels.shape
		except tf.errors.OutOfRangeError:
			print "Done training -- epoch limit reached"
		finally:
			coord.request_stop()

		coord.join(threads)

**Comment by chenghuige** [comment]
@ebrevdo  dose  tf.parse_single_sequence_example  works with new input pipeline api dataset ?

**Comment by ebrevdo** [comment]
Yes it does.
